pkg/cluster/executor/ssh.go

```diff
@@ -207,3 +226,111 @@ func (e *SSHExecutor) Transfer(src string, dst string, download bool) error {
+	if e.Sudo || sudo {
+		cmd = fmt.Sprintf("sudo -H -u root bash -c \"%s\"", cmd)
+	}
+
+	// set a basic PATH in case it's empty on login
+	cmd = fmt.Sprintf("PATH=$PATH:/usr/bin:/usr/sbin %s", cmd)
+
+	if e.Locale != "" {
+		cmd = fmt.Sprintf("export LANG=%s; %s", e.Locale, cmd)
+	}
+
+	// run command on remote host
+	// default timeout is 60s in easyssh-proxy
+	if len(timeout) == 0 {
+		timeout = append(timeout, executeDefaultTimeout)
+	}
+
+	ctx := context.Background()
+	if len(timeout) > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(context.Background(), timeout[0])
+		defer cancel()
+	}
+
+	args := []string{"ssh", "-o", "StrictHostKeyChecking=no"}
+	if e.Config.Password != "" {
+		args = append([]string{"sshpass", "-p", e.Config.Password, "-P", "password"}, args...)
+	}
+	if e.Config.KeyFile != "" {
+		args = append(args, "-i", e.Config.KeyFile)
+	}
+	if e.Config.Passphrase != "" {
+		args = append([]string{"sshpass", "-p", e.Config.Passphrase, "-P", "passphrase"}, args...)
+	}
+	args = append(args, fmt.Sprintf("%s@%s", e.Config.User, e.Config.Host), cmd)
+
+	command := exec.CommandContext(ctx, args[0], args[1:]...)
+
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+	command.Stdout = stdout
+	command.Stderr = stderr
+	err := command.Run()
+
+	zap.L().Info("SSHCommand",
+		zap.String("host", e.Config.Host),
+		zap.Int("port", e.Config.Port),
+		zap.String("cmd", cmd),
+		zap.Error(err),
+		zap.String("stdout", stdout.String()),
+		zap.String("stderr", stderr.String()))
+
+	if err != nil {
+		baseErr := ErrSSHExecuteFailed.
+			Wrap(err, "Failed to execute command over SSH for '%s@%s:%d'", e.Config.User, e.Config.Host, e.Config.Port).
+			WithProperty(ErrPropSSHCommand, cmd).
+			WithProperty(ErrPropSSHStdout, stdout).
+			WithProperty(ErrPropSSHStderr, stderr)
+		if len(stdout.Bytes()) > 0 || len(stderr.Bytes()) > 0 {
+			output := strings.TrimSpace(strings.Join([]string{stdout.String(), stderr.String()}, "\n"))
+			baseErr = baseErr.
+				WithProperty(cliutil.SuggestionFromFormat("Command output on remote host %s:\n%s\n",
+					e.Config.Host,
+					color.YellowString(output)))
+		}
+		return stdout.Bytes(), stderr.Bytes(), baseErr
+	}
+
+	return stdout.Bytes(), stderr.Bytes(), err
+}
+
+// Transfer copies files via SCP
+// This function depends on `scp` (a tool from OpenSSH or other SSH implementation)
+// This function is based on easyssh.MakeConfig.Scp() but with support of copying
+// file from remote to local.
+func (e *NativeSSHExecutor) Transfer(src string, dst string, download bool) error {
+	args := []string{"scp", "-r", "-o", "StrictHostKeyChecking=no"}
+	if e.Config.Password != "" {
+		args = append([]string{"sshpass", "-p", e.Config.Password, "-P", "password"}, args...)
+	}
+	if e.Config.KeyFile != "" {
+		args = append(args, "-i", e.Config.KeyFile)
+	}
+	if e.Config.Passphrase != "" {
+		args = append([]string{"sshpass", "-p", e.Config.Passphrase, "-P", "passphrase"}, args...)
```

Comment:
Are you sure the `sshpass` always exists in the user environment?

Reply:
No, It's a tradeoff. I think it's not a hard task to install it in the bastion host.